### PR TITLE
Fix harness in lib section

### LINKED ISSUE #2305: "Could not execute process (never executed)" with [lib] harness = false
http://doc.crates.io/manifest.html#configuring-a-target suggests that `harness = false` can be used in the `[lib]` section of `Cargo.toml`. Let’s try it:

```rust
/tmp% cargo new foo
/tmp% cd foo 
/tmp/foo% echo -e '[lib]\nharness=false' >> Cargo.toml
/tmp/foo% cargo test -v
   Compiling foo v0.1.0 (file:///tmp/foo)
     Running `rustc src/lib.rs --crate-name foo --crate-type lib -g -C metadata=9476bb3dbe366490 -C extra-filename=-9476bb3dbe366490 --out-dir /tmp/foo/target/debug --emit=dep-info,link -L dependency=/tmp/foo/target/debug -L dependency=/tmp/foo/target/debug/deps`
     Running `rustc src/lib.rs --crate-name foo --crate-type lib -g --out-dir /tmp/foo/target/debug --emit=dep-info,link -L dependency=/tmp/foo/target/debug -L dependency=/tmp/foo/target/debug/deps`
     Running `/tmp/foo/target/debug/foo-9476bb3dbe366490`
Could not execute process `/tmp/foo/target/debug/foo-9476bb3dbe366490` (never executed)

Caused by:
  Could not execute process `/tmp/foo/target/debug/foo-9476bb3dbe366490` (never executed)

Caused by:
  No such file or directory (os error 2)
```

`--test` is indeed not passed, but `--crate-type lib` still is so no executable is created.

Grepping Cargo’s source shows an example in `tests/test_cargo_test.rs` where `harness = false`  is used in `[[test]]`, which makes more sense.

So:

* The docs should show `harness = false` in a `[[test]]` section, not `[lib]`.
* `harness = false` in `[lib]` should probably emit a warning and be otherwise ignored, since it doesn’t work. Or make Cargo emit in error and abort.
* With `harness = false` in `[[bin]]`, there is no way that I can find to tell whether an executable is being built for testing or not. Maybe pass `--cfg test`?

### PR DESCRIPTION
**Pull request by TheNeikos**
closes #2305

### CODE REVIEW

**Comment by TheNeikos** [comment]
I have to admit I made this somewhat 'blindly' as I did not find a way to correctly test the part where I pass the cfg.

If someone can point me to a way to do it. I will try to think of something too.

**Comment by alexcrichton** [comment]
Thanks! You should be able to add a test with a `harness = false` and then something like:

```rust
#[cfg(test)]
fn main() {
}
```

(that'll fail to compile if `#[cfg(test)]` isn't passed)

File: src/cargo/core/manifest.rs [hunk 1]
```diff
@@ -384,7 +384,10 @@ impl Target {
         self
     }
     pub fn set_harness(&mut self, harness: bool) -> &mut Target {
-        self.harness = harness;
+        match  self.kind {
+            TargetKind::Lib(_) => (),
+            _ => self.harness = harness,
+        }
```
Comment (alexcrichton) [inline src/cargo/core/manifest.rs]: Hm I may be forgetting now, but this should be ok, right? If we pass `--cfg test` then a library can have its own custom test harness/main function, so it seems legitimate that `harness = false` could work for a library?

File: tests/test_cargo_test.rs [hunk 2]
```diff
@@ -1206,6 +1206,46 @@ test!(test_no_harness {
...
+            [package]
+            name = "foo"
+            version = "0.0.1"
+            authors = []
+
+            [lib]
+            name = "foo"
+            harness = false
+        "#)
+        .file("src/lib.rs", "#[cfg(test)] fn main() {}");
```
Comment (TheNeikos) [inline tests/test_cargo_test.rs]: @alexcrichton Is this how you meant it? I understood it as being the entrypoint of the test harness in this case. 
However `#[cfg(foo)]` also passes. So I am a  bit confused.

Comment (alexcrichton) [inline tests/test_cargo_test.rs]: Ah yeah you'll want to remove the hardcoded check I commented on above, then it should fail for `#[cfg(foo)]` and this should pass.

Comment (TheNeikos) [inline tests/test_cargo_test.rs]: Restoring the old functionality of the function you commented about still makes it pass with `#[cfg(foo)]`

File: src/cargo/core/manifest.rs (hunk 1 above)
Comment (TheNeikos) [inline src/cargo/core/manifest.rs]: If `harness == false` then there is no binary produced. I have not been able to find out why that is, but the error goes similar to the linked issue.

**Comment by alexcrichton** [comment]
Perhaps `--lib` is still getting passes somewhere? It may need to be dropped in the case that a test profile is in play

File: src/doc/manifest.md [hunk 3]
```diff
@@ -512,9 +512,10 @@ doc = true
 # for Cargo to correctly compile it and make it available for all dependencies.
 plugin = false
 
+[[test]]
 # If set to false, `cargo test` will omit the --test flag to rustc, which stops
 # it from generating a test harness. This is useful when the binary being built
-# manages the test runner itself.
+# manages the test runner itself. It will still pass the cfg feature 'test'.
```
Comment (IvanUkhov) [inline src/doc/manifest.md]: If possible, please change the straight quotes around _test_ to backticks.

**Comment by alexcrichton** [comment]
I did some investigation to try to land this, but it's unfortunately blocked by https://github.com/rust-lang/rust/issues/33670, so I'm going to close this for now. Thanks though for the PR @TheNeikos!

**Pull request closed without merging**